src/Streamly/Data/Unfold.hs

```diff
@@ -8,63 +8,106 @@
+-- 5050
+--
+-- All the serial stream generation operations in "Streamly.Prelude"
+-- can be expressed using unfolds:
+--
+-- > Stream.fromList = Stream.unfold Unfold.fromList [1..100]
+--
+-- Conceptually, an 'Unfold' is just like "Data.List.unfoldr". Let us write a
+-- step function to unfold a list using "Data.List.unfoldr":
+--
+-- >>> :{
+--  f [] = Nothing
+--  f (x:xs) = Just (x, xs)
+-- :}
+--
+-- >>> Data.List.unfoldr f [1,2,3]
+-- [1,2,3]
+--
+-- Unfold.unfoldr is just the same, it uses the same step function:
+--
+-- >>> Stream.toList $ Stream.unfold (Unfold.unfoldr f) [1,2,3]
+-- [1,2,3]
+--
+-- The input of an unfold can be transformed using 'lmap':
+--
+-- >>> u = Unfold.lmap (fmap (+1)) Unfold.fromList
+-- >>> Stream.toList $ Stream.unfold u [1..5]
+-- [2,3,4,5,6]
 --
+-- 'Unfold' streams can be transformed using transformation combinators. For
+-- example, to retain only the first two elements of an unfold:
+--
+-- >>> u = Unfold.take 2 Unfold.fromList
+-- >>> Stream.toList $ Stream.unfold u [1..100]
+-- [1,2]
+--
+-- Multiple unfolds can be combined in several interesting ways. For example,
+-- to generate nested looping as in imperative languages (also known as cross
+-- product of the two streams):
+--
+-- >>> u1 = Unfold.lmap fst Unfold.fromList
+-- >>> u2 = Unfold.lmap snd Unfold.fromList
+-- >>> u = Unfold.crossWith (,) u1 u2
+-- >>> Stream.toList $ Stream.unfold u ([1,2,3], [4,5,6])
+-- [(1,4),(1,5),(1,6),(2,4),(2,5),(2,6),(3,4),(3,5),(3,6)]
+--
+-- Nested loops using unfolds provide C like performance due to full stream
+-- fusion.
+--
+-- Please see "Streamly.Internal.Data.Unfold" for additional @Pre-release@
+-- functions.
+--
+-- = Unfolds vs. Streams
+--
+-- We can often use streams or unfolds to achieve the same goal. In general,
+-- all the serial stream generation and transformation primitives in
+-- "Streamly.Prelude" can be expressed using unfolds. However, for streams,
+-- nested looping operations using the 'Monad' instance or
+-- 'Streamly.Prelude.concatMap' cannot fuse completely.
+-- 'Streamly.Prelude.unfoldMany', 'many' and 'crossWith' operations are high
+-- performance alternatives to @concatMap@ based nested loop operations.
+--
+-- While unfold nesting operations provide the best performance the stream
+-- concatMap or Monad instance operations provide more power. However, that
+-- power may not always be needed. Also, as of now unfolds do not support
+-- concurrent composition whereas streams do. The
+-- 'Streamly.Prelude.concatMapWith' operation in streams can provide powerful
+-- concurrent composition semantics.
+--
+-- To summarize, the only place where you would need unfolds over streams is
+-- serial nested operations to gain performance via full stream fusion.
+-- However, most of the serial stream generation operations in streamly are
+-- implemented as unfolds so that they can be used efficiently with nested
+-- operations and can be converted to streams using Stream.unfold when needed.
```

Comment:
```suggestion
-- However, most of the serial stream generation operations in streamly are
-- implemented as unfolds so that they can be used efficiently with nested
-- operations and can be converted to streams using 'Stream.unfold' when needed.
```